### PR TITLE
Fix test_matmul_mv_cpu_float32 and Enable test_linalg for AArch64

### PR DESCRIPTION
**Pull request by robert-hardwick**
I have edited this PR multiple times.

### Problem Summary ###

The problem we are trying to solve is that we see test_matmul_mv_cpu_float32 and test_matmul_mv_cpu_float16 and test_matmul_mv_cpu_bfloat16 fail for various reasons on AArch64. 

I have done a thorough investigation and some of these tests are passing simply by pure random chance. test_matmul_mv_cpu_bfloat16 passes in it's current form but can be made to fail simply by changing the random seed.

test_matmul_mv_cpu_float32 fails but equally can be made to pass by setting the random seed.

The failure message is this something like this

```
Mismatched elements: 50000 / 50000 (100.0%)
Greatest absolute difference: 0.0859375 at index (0,) (up to 1e-05 allowed)
Greatest relative difference: 3.4299155231565237e-06 at index (0,) (up to 1.3e-06 allowed)

To execute this test, run the following from the base repo dir:
    python test/test_linalg.py TestLinalgCPU.test_matmul_mv_cpu_float32 
```

The test is the following

```
        n = 50_000
        A = torch.ones(n, n, dtype=dtype, device=device)
        B = torch.rand(n, dtype=dtype, device=device)
        C = torch.matmul(A, B)
        
        self.assertEqual(C, B.sum().expand(B.shape))
```

The problem is that we are testing 2 different reduction paths for a very large N( torch.matmul and torch.sum will perform a different order of summation, especially with vectorization  ). So it is entirely expected that C and B.sum() will have large differences.

This is made worse by the use of the uniform distribution instead of the normal distribution. Below are 2 charts which demonstrate that errors tends to accumulate into larger magnitudes for the uniform distribution. ( Note the x axis is square root of N )

### mean error calculated over different 100 seeds for torch.rand

linalg_matmul

### mean error calculated over different 100 seeds for torch.randn

randn

There is some mathematical proof for this.... following a random walk model to calculate the expected error .. with normally distributed data the error scales with square root of N ( N^1/2 ) , where as for uniformly distributed data the error scales with N^3/2 .

The reason being is that for normally distributed data the running sum tends to oscillate around 0, so the magnitude of the error stays roughly constant, where as for non-zero mean random data the sum grows and so does the magnitude of the error. 

### The Fix ###

- I have changed rand to randn. 
- Technically the comment says this is regression test for a segmentation fault https://github.com/pytorch/pytorch/issues/150637 so in theory we don't need to check for closeness, we could just put sanity checks in opn the size and shape.
- For completeness we keep the assertEqual but i have overridden the tolerances with suitably large atol and rtol which have been made from the mathematical theory but also from evidence of running this with 100 seeds.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #167069

### CODE REVIEW

**Review by malfet — APPROVED** [review]
LGTM if CI is green (but it is not)

File: test/test_linalg.py [hunk 1]
```diff
@@ -10002,7 +10002,18 @@ def test_matmul_mv(self, device, dtype):
...
+
+        # due to large N ( 50_000 ), accumulation and assosiativity errors can occur
+        # therefore we calculate appropriate rtol and atol based on N and machine epsilon.
+
+        b_sum_f32 = B.float().sum()  # compute sum, ensure f32 accumulation
+        safety_factor = 3
+        eps = torch.finfo(dtype).eps
+        u = eps / 2.0         # unit roundoff for dtype
+        rtol = min(5e-2, max(eps, safety_factor * u * math.sqrt(n)))   # ~ O(u * sqrt(N)) with small safety factor
+        atol = rtol * b_sum_f32.abs().item()
```
Comment (jondea) [inline test/test_linalg.py]: This is just rescaling the atol to be an rtol. It would be clearer if we just didn't do it at all.

File: test/test_linalg.py [hunk 2]
```diff
@@ -10002,7 +10002,18 @@ def test_matmul_mv(self, device, dtype):
...
-        self.assertEqual(C, B.sum().expand(B.shape))
+
+        # due to large N ( 50_000 ), accumulation and assosiativity errors can occur
+        # therefore we calculate appropriate rtol and atol based on N and machine epsilon.
+
+        b_sum_f32 = B.float().sum()  # compute sum, ensure f32 accumulation
+        safety_factor = 3
+        eps = torch.finfo(dtype).eps
+        u = eps / 2.0         # unit roundoff for dtype
+        rtol = min(5e-2, max(eps, safety_factor * u * math.sqrt(n)))   # ~ O(u * sqrt(N)) with small safety factor
```
Comment (jondea) [inline test/test_linalg.py]: Why `5e-2`?

File: test/test_linalg.py [hunk 3]
```diff
@@ -10002,7 +10002,18 @@ def test_matmul_mv(self, device, dtype):
         A = torch.ones(n, n, dtype=dtype, device=device)
         B = torch.rand(n, dtype=dtype, device=device)
         C = torch.matmul(A, B)
-        self.assertEqual(C, B.sum().expand(B.shape))
+
+        # due to large N ( 50_000 ), accumulation and assosiativity errors can occur
+        # therefore we calculate appropriate rtol and atol based on N and machine epsilon.
+
+        b_sum_f32 = B.float().sum()  # compute sum, ensure f32 accumulation
```
Comment (jondea) [inline test/test_linalg.py]: What about when dtype == f32? Can we do f64?

File: test/test_linalg.py (hunk 2 above)
Comment (jondea) [inline test/test_linalg.py]: I'm not convinced by this formula. I think the accumulation type is f32, so the `u` in this case should be `eps(f32)`. Then you need to account for quantization error, which is always `eps(bf16)/2`. Also, `sqrt` makes sense for normally distributed inputs, where the round off error will be random walk. But for uniform inputs, I'm not convinced.

I'm a little bit confused why there is no common way that PyTorch handles checking if numerical error is acceptable.

**Review by jondea — CHANGES_REQUESTED** [review]

**Comment by jondea** [comment]
Broadly speaking, I agree that these numerical errors look fine, I just disagree with your formula. More generally, I don't think it makes sense to check the absolute error of a large floating point number at all.

File: test/test_linalg.py (hunk 2 above)
Comment (robert-hardwick) [reply test/test_linalg.py]: this is just a cap to stop rtol from exploding to absurdly large numbers when n is large and precision is low e.g. bfloat16

File: test/test_linalg.py (hunk 1 above)
Comment (robert-hardwick) [reply test/test_linalg.py]: Yes, that's deliberate. In this is particular to this test where every element in `B.sum().expand(B.shape)` is the magnitude.

It makes the absolute tolerance correspond to the same proportional error allowed by the relative tolerance

File: test/test_linalg.py (hunk 2 above)
Comment (jondea) [reply test/test_linalg.py]: If the actual error is expected to explode, how can we stop it? Also, I think this bound is passed at n=19

```
3*eps(BFloat16)*sqrt(19)/2 = 0.051
```

File: test/test_linalg.py (hunk 1 above)
Comment (jondea) [reply test/test_linalg.py]: Indeed, but this atol is not doing anything. It would be clearer to just not test the atol at all and rely on rtol. The intent would then be clear.

File: test/test_linalg.py (hunk 2 above)
Comment (robert-hardwick) [reply test/test_linalg.py]: apologies, i wasn't clear

The cap isn’t just about “large n”; it’s about keeping rtol realistic whenever the result of the formula gets too big.... and that can happen either because n is large or because the dtype is coarse.

I think n=19 is unique to bf16, in float32 it's much much larger.... apparently in the billions??

File: test/test_linalg.py (hunk 1 above)
Comment (robert-hardwick) [reply test/test_linalg.py]: Closing this thread as the fix for this test has changed so we no longer need to change the tolerances.

File: test/test_linalg.py (hunk 2 above)
Comment (robert-hardwick) [reply test/test_linalg.py]: As above.
You are correct that uniform inputs changes the relationship between N and the expected error, which lead me to fix this by changing the random input to be normally distrubuted.

**Review by jondea — APPROVED** [review]
That's an even better fix, thanks!

File: test/test_linalg.py [hunk 4]
```diff
@@ -17,6 +17,10 @@
 from typing import Union, Optional
 from torch._prims_common import DimsType
 from packaging import version
+from torch.testing._internal.common_device_type import (
+    tol,
```
Comment (jondea) [inline test/test_linalg.py:21]: You could import as or just use it fully qualified to avoid the name clash.

**Review by jondea — APPROVED** [review]
One suggestion, otherwise looks good
